keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,389 @@
+one way of handling a particular Job failure reason.
+
+- OnJobFailureReasons must be a valid Job failure reason as defined [here](https://github.com/kubernetes/kubernetes/blob/2d4100335e4c4ccc28f96fac78153f378212da4c/staging/src/k8s.io/api/batch/v1/types.go#L537-L554)
+in the batchv1 Job API. At the time of writing this KEP, these include:
+  - `PodFailurePolicy`
+  - `BackoffLimitExceeded`
+  - `DeadlineExceeded`
+  - `MaxFailedIndexesExceeded`
+  - `FailedIndexes`
+
+### Implmentation
+
+The core part of the implementation will be defining what specific mechanisms the JobSet controller uses to implement
+the behavior defined for each FailurePolicyAction type:
+
+1) `RestartJob`: To restart a single child job without restarting the entire JobSet, the controller will delete that 
+particular child job and allow the normal reconciliation process to recreate it.
+
+2) `FailJob`: To leave a particular child job in a failed state without restarting it or restarting the JobSet, the
+controller will simply do nothing, taking no action on this job.
+
+3) `FailJobSet`: To fail the JobSet immediately without restarting, the controller updates the JobSet status to failed.
+
+4) `Ignore`: To "ignore" the failure (i.e., restart the JobSet without counting it against `MaxRestarts`), the controller
+will delete all child jobs and allow the normal reconciliation process to recreate them. This is in contrast to how
+we restart the JobSet **without** ignoring the failure, which is done by incrementing the counter on the JobSet annotation
+`jobset.sigs.k8s.io/restart-attempt`, which on subsequent reconiliations will trigger job deletion and recreation for any
+child job with a `jobset.sigs.k8s.io/restart-attempt` counter value less than that of the JobSet.
+
+
+### Test Plan
+
+The testing plan will focus on integration tests. Specific test cases and scenarios are defined in the integration test
+section below.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+#### Unit Tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
+
+#### Integration tests
```

Comment:
What is our policy for e2e tests?

@ahg-g @danielvegamyhre 

It could be nice to start adding some e2e tests for features like this. But we don't have a lot of coverage of e2e tests at the moment.

Reply:
In general, our informal/de fact policy for e2e tests has been only adding e2e tests for things which cannot be tested via unit tests or integration tests. The pod networking test cases are an example of something which can only be verified via an e2e test, since we need the Job controller to be running and to actually create pods. 

In this case, we don't care about pods, just about how the JobSet controller interacts with child jobs based on the failure policy, which is something we can test with integration tests.